pkg/provisioner/ironic/ironic.go

```diff
@@ -1295,3 +1327,46 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// softPowerOff sends 'soft power off' request to BM node.
+// If soft power off is not supported, the request ends with an error.
+// Otherwise the request ends with no error and the result should be
+// checked later via node fields "power_state" and "target_power_state".
+func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error) {
+	p.log.Info("ensuring host is powered off by \"soft power off\" command")
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+
+	if ironicNode.PowerState != powerOff {
+		targetState := ironicNode.TargetPowerState
+		// If the target state is either powerOff or softPowerOff, then we should wait
+		if targetState == powerOff || targetState == softPowerOff {
+			p.log.Info("waiting for power status to change")
+			result.RequeueAfter = powerRequeueDelay
+			result.Dirty = true
+			return result, nil
+		}
+		// If the target state is unset, then soft power off has failed
+		if p.host.Status.SoftPowerOffTriggered && targetState == "" {
```

Comment:
Can we check the target_state + the last_error to distinguish between the cases where we haven't started a power off vs. a previous attempt has timed out, instead of having to set Status.SoftPowerOffTriggered?

Reply:
I've tried your suggestion. It seems fine for the common cases I test. However, I think using the LastError may not be reliable as using a "trigger" flag.